### PR TITLE
ref(v2): refactor state management in Create Tab

### PR DESCRIPTION
**Pull request by mantariksh**
## Problem

The existing form builder has a couple of bugs:
- After creating a new field, the onSuccess handler in the Edit* component does not fire, because the component unmounts and re-mounts due to its key in the parent changing.
- After updating a field and exiting the edit field drawer, the field reverts back to its previous state. When attempting to create a new field, the field flips into the new state again.

I found it difficult to address these bugs because I found the state management code difficult to follow (with due kudos to @karrui who implemented it amazingly quickly for Hackathon). In particular, I found the concept of an `activeField` (which could be either a field being created or an existing field being edited) as well as a `fieldToCreate` (containing only a field type and an index) confusing because I wasn't sure what states they mapped to in the UI (when are both supposed to be defined vs one vs the other? does one take precedence? etc).

I also felt that determining whether the current field is an existing or new field by checking whether its `_id` is a hardcoded string was slightly janky and implicit (again with due deference to @karrui for implementing it super fast). Given that this distinction is of central importance to the Create tab (it determines what API calls should be made, and what happens if the user closes the drawer before saving), I thought we could be more explicit about it.

As a result of my confusion, I thought that we had to overhaul the state management in order to address the above bugs. However, I realised as I was writing this code that there might have been simpler fixes. Nevertheless, I hope the new code is more understandable and maintainable.

Also in hindsight, I'm not sure whether I found it confusing because of the code itself or because the documentation could have been better. It's entirely possible that my code is equally confusing to other readers, in which case I'd be happy to revert back to the previous version and try to document it better.

Anyway...

## Solution

Refactor the state management such that the store contains a `stateData` object, which can be in the following states:
- `CreatingField`
  - Includes a `field` which is the current state of the field, and an `insertionIndex`
- `EditingField`
  - Includes a `field` which is the current state of the field
- `Inactive`

More states can be added in future, e.g. for editing start and end pages.

This allows us to be explicit about the current state of the Create tab instead of implicitly storing the state in the field `_id`. The `PENDING_CREATE_FIELD_ID` still exists but merely as an identifier for `react-hook-form`.

Also moves around and renames a bunch of files and components in a way that made more sense to me. Sorry. For example, I kept thinking that `CreateFieldDrawer` was a component for modifying a field which was going to be created, but actually it referred to the list of fields which could be created, so I renamed it to `FieldListDrawer`.

## Other improvements
- Added an MSW mock for reordering fields

## Known issues
- The previous version had a checkbox editing drawer, but I haven't included it here as I'd like to get this approach reviewed before doing that. No point doing it now since we might choose to stick to the existing state management.
- ~The `EditFieldDrawer` renders twice when it mounts, and I haven't been able to figure out why (before I go on my break next week). Please feel free to edit this branch to fix this if you find the problem.~ resolved in [c9740ca](https://github.com/opengovsg/FormSG/pull/3547/commits/c9740ca1c5f29b2cabcaf149f74611bf3de80b5d)

## Future work
- I think it might be a good idea to add stories where a particular API call always fails, so we can test the error states.

### CODE REVIEW

**Comment by mantariksh** [comment]
please feel free to modify this branch while I'm away!

File: frontend/src/features/admin-form/create/builder-and-design/dndHelpers.ts
Comment (karrui) [inline frontend/src/features/admin-form/create/builder-and-design/dndHelpers.ts:1]: nit: should we slot this under `builder-and-design/utils/dnd.ts` instead? (and move the existing `utils.ts` into `utils/fieldCreation.ts` instead

File: frontend/src/features/admin-form/create/builder-and-design/useCreateTabForm.ts
Comment (karrui) [inline frontend/src/features/admin-form/create/builder-and-design/useCreateTabForm.ts:13]: can be changed to 
```ts
enabled: stateData === null
```
if we used null to represent inactivity instead.

File: frontend/src/features/admin-form/create/common/CreatePageSidebarContext/CreatePageSidebarContext.tsx
Comment (karrui) [inline frontend/src/features/admin-form/create/common/CreatePageSidebarContext/CreatePageSidebarContext.tsx:60]: Rereading this; maybe we should put these invocations in a `useEffect` whenever active tab changes instead.

```tsx
useEffect(() => {
  setFieldsToInactive()
}, [activeTab])
```

**Review by karrui — COMMENTED** [review]
Some comments.

RE: rerendering: We can actually trace it (if you really want) with https://www.npmjs.com/package/@welldone-software/why-did-you-render

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/BuilderAndDesignContent.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/BuilderAndDesignContent.tsx]: forgot to remove this, actually not required for `BuilderFields` component. 9694f408

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/FieldRow/FieldRowContainer.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/FieldRow/FieldRowContainer.tsx:53]: ooo imo existing version is more readable haha so will keep it as-is!

File: frontend/src/features/admin-form/create/builder-and-design/useBuilderAndDesignStore.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/useBuilderAndDesignStore.tsx:29]: yeap that was the initial implementation but I changed it to add an inactive state to make it more explicit. I think even if we allowed the state to be `null`, we would still need specific checks like the first one in your comment because there will likely be more states in future (`EditingStartPage` and `EditingEndPage`). so personally I'd rather keep it explicit to force these checks. wdyt?

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/BuilderAndDesignPlaceholder/BuilderAndDesignPlaceholder.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/BuilderAndDesignPlaceholder/BuilderAndDesignPlaceholder.tsx:1]: thanks! leftover from debugging

File: frontend/src/features/admin-form/create/builder-and-design/useBuilderAndDesignStore.tsx
Comment (karrui) [reply frontend/src/features/admin-form/create/builder-and-design/useBuilderAndDesignStore.tsx:29]: I'll defer to you!

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/BuilderAndDesignPlaceholder/BuilderAndDesignPlaceholder.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/BuilderAndDesignPlaceholder/BuilderAndDesignPlaceholder.tsx]: ec240153

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignTab.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignTab.tsx:41]: good point! wanted to remove this but I forgot

Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignTab.tsx]: af940c4d

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/useBuilderFields.ts
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/useBuilderFields.ts]: af940c4d

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/EditFieldDrawer.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/EditFieldDrawer.tsx:41]: nope unfortunately

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/BuilderAndDesignDrawer.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/BuilderAndDesignDrawer.tsx:49]: keeping inactive state as-is as per prev comment

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/EditFieldDrawer.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/EditFieldDrawer.tsx]: 85771b41

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/EditHeader.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/EditHeader.tsx:56]: I don't think so actually. if we unmount when the fieldId changes then the component will unmount and remount when a newly created field is changed, because the fieldId will change from the pending ID to the field's actual ID. then the `onSuccess` or `onFailure` handlers for the don't fire because the component is unmounted when the mutation completes.

File: frontend/src/features/admin-form/create/builder-and-design/useCreateTabForm.ts
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/useCreateTabForm.ts:13]: keeping state management as-is as per previous comment

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/EditHeader.tsx
Comment (karrui) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/EditHeader.tsx:56]: Hmm, can we update the onSuccess handler of the createField query to update the fieldId instead of doing it earlier? That would allow the onSuccess to run before the id is updated

File: frontend/src/features/admin-form/create/builder-and-design/dndHelpers.ts
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/dndHelpers.ts]: thanks! 69c4e9bb

File: frontend/src/features/admin-form/create/common/CreatePageSidebarContext/CreatePageSidebarContext.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/common/CreatePageSidebarContext/CreatePageSidebarContext.tsx]: 8e8eadb4

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/EditHeader.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/EditHeader.tsx:56]: but still a bit janky right - if we add an `onSuccess` on the component in future, it will fail

File: frontend/src/features/admin-form/create/common/CreatePageSidebarContext/CreatePageSidebarContext.tsx
Comment (timotheeg) [inline frontend/src/features/admin-form/create/common/CreatePageSidebarContext/CreatePageSidebarContext.tsx:72]: (react noob question) Is using hooks the best way to achieve this? `setActiveTab` is guaranteed to be stable across invocations, which mean `handleBuilderClick`, `handleDesignClick`, and `handleLogicClick` are also guaranteed to be stable references across invocations, and so in all these `useCallback()` calls, we create local functions which are just discarded. This looks a lot more like construction-time concern rather than render-time concern.

Comment (timotheeg) [reply frontend/src/features/admin-form/create/common/CreatePageSidebarContext/CreatePageSidebarContext.tsx:72]: Hm, nvm. [This answer](https://reactjs.org/docs/hooks-faq.html#are-hooks-slow-because-of-creating-functions-in-render) indicates that in practice we just shouldn't have to worry about it. It will take me some time to get use to this idea 😅

**Review by timotheeg — COMMENTED** [review]
Thanks for taking the time to do this Anty. I think the approach is sane 👍 .

Moving the state into the drawer and using explicit states does help with readability, although just to note on one tradeoff we are making by doing that: there are now conditional checks with no default `else`, which "look" like some control flow could be missing (e.g. in `EditFieldDrawer`'s `handleSave()` and `handleChange()`). Of course nothing is missing because all cases are accounted for, but the code itself is not reflecting that.

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/useBuilderFields.ts
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignContent/useBuilderFields.ts]: thanks tim! yup this was leftover from a previous implementation. fixed in d208ad76

File: frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/EditHeader.tsx
Comment (mantariksh) [reply frontend/src/features/admin-form/create/builder-and-design/BuilderAndDesignDrawer/EditFieldDrawer/edit-fieldtype/EditHeader.tsx]: thanks tim, this was indeed a bug with variable naming which I fixed without realising 😅 

before this refactor, `isSaveDisabled` was passed to `FormFieldDrawerActions` as the `isDirty` prop (`isDirty={isSaveDisabled}`, which was then used in the `Button` as `isDisabled={!isDirty}` - which means the `isSaveDisabled` variable was actually for whether saving was enabled.

previously I fixed this by flipping the conditional inside the `useMemo`, but I've now taken your suggestion and changed the variable name to `isSaveEnabled` in 7bc1ec38.

**Review by karrui — APPROVED** [review]
lgtm, i see you tracked additional render reason 😍

**Comment by karrui** [comment]
This didn't pass design review before merge; mobile views do not render the content. I'll open a PR to fix it.

**Pull request merged**
